apps/client/src/common/hooks/useEventAction.ts

```diff
@@ -316,5 +324,86 @@ export const useEventAction = () => {
     [_reorderEventMutation],
   );
 
-  return { addEvent, updateEvent, deleteEvent, deleteAllEvents, applyDelay, reorderEvent };
+  /**
+   * Calls mutation to swap events
+   * @private
+   */
+  const _swapEvents = useMutation({
+    mutationFn: requestEventSwap,
+    // we optimistically update here
+    onMutate: async ({ from, to }) => {
+      // cancel ongoing queries
+      await queryClient.cancelQueries(RUNDOWN_TABLE, { exact: true });
+
+      // Snapshot the previous value
+      const previousEvents = [...(queryClient.getQueryData(RUNDOWN_TABLE) as OntimeRundown)];
+
+      const fromEventIndex = previousEvents.findIndex((event) => event.id === from);
+      const toEventIndex = previousEvents.findIndex((event) => event.id === to);
+
+      const fromEvent = previousEvents.at(fromEventIndex) as OntimeEvent | undefined;
+      const toEvent = previousEvents.at(toEventIndex) as OntimeEvent | undefined;
+
+      if (!fromEvent || !toEvent) {
+        return { previousEvents };
+      }
+
+      previousEvents[fromEventIndex] = {
+        ...toEvent,
+        timeStart: fromEvent.timeStart,
+        timeEnd: fromEvent.timeEnd,
+        duration: fromEvent.duration,
+      };
+
+      previousEvents[toEventIndex] = {
+        ...fromEvent,
+        timeStart: toEvent.timeStart,
+        timeEnd: toEvent.timeEnd,
+        duration: toEvent.duration,
+      };
+
```

Comment:
I got confused when I replied earlier and gave the wrong answer, we need to add the delay here

``` ts
      previousEvents[fromEventIndex] = {
        ...toEvent,
        timeStart: fromEvent.timeStart,
        timeEnd: fromEvent.timeEnd,
        duration: fromEvent.duration, 
        delay: fromEvent.delay
      };

      previousEvents[toEventIndex] = {
        ...fromEvent,
        timeStart: toEvent.timeStart,
        timeEnd: toEvent.timeEnd,
        duration: toEvent.duration,
        delay: toEvent.delay
      };
```

Reply:
I see. However, `delay` seems to be an optional property. If we add it and it's not present, we risk attaching `delay: undefined` to the property, right?